### PR TITLE
Remove sh dependency

### LINKED ISSUE #6547: Remove `sh` dependency
*This is a good first issue for new contributors to take on, if you have any questions, please ask on the task or in our [Gitter room](https://gitter.im/freedomofpress/securedrop)!*

## Description

`sh` is a pretty cool project, it allows you to shell out to programs as if they were native Python programs. This comes with a downside, it obfuscates the underlying command being run, and it's an extra dependency.

Here are all the places we use it:
```
Targets
    Occurrences of 'from sh import' in Project
Found Occurrences in Project  (4 usages found)
    Unclassified  (4 usages found)
        securedrop  (4 usages found)
            securedrop  (1 usage found)
                i18n_tool.py  (1 usage found)
                    19 from sh import git, msgfmt, msgmerge, pybabel, sed, xgettext
            securedrop/tests  (3 usages found)
                test_i18n.py  (1 usage found)
                    34 from sh import pybabel, sed
                test_i18n_tool.py  (1 usage found)
                    13 from sh import git, msginit, pybabel, sed, touch
                test_template_filters.py  (1 usage found)
                    13 from sh import pybabel
```

Given how minimal its usage is (just `i18n_tool.py` plus tests), I propose we replace it with explicit `subprocess.check_output(...)` calls. The biggest advantage of this removal is that `pybabel` will be the only dependency needed to compile translations at package build time.

### PR DESCRIPTION
**Pull request by lb803**
## Status

Ready for review

## Description of Changes

Fixes #6547.

I removed all the `sh` dependencies with the exception of `sh.sed`, as this has been addressed in another contribution (https://github.com/freedomofpress/securedrop/issues/6547#issuecomment-1256930073).

## Testing

Small changes, limited to two files:
$ `securedrop/bin/dev-shell bin/run-test tests/test_i18n_tool.py`
$ `securedrop/bin/dev-shell bin/run-test tests/test_i18n.py`

## Deployment

Ready

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container
- [x] These changes do not require documentation

### CODE REVIEW

File: securedrop/i18n_tool.py [hunk 1]
```diff
@@ -427,7 +450,8 @@ def commit_changes(
           commit: {current}
         """
         ).format(remote=args.url, name=name, authors=authors_as_str, code=code, current=current)
-        git("-C", args.root, "commit", "-m", message, *paths)
+        subprocess.run(["git", "commit", "-m", message, *paths], cwd=args.root)
+        log.debug(f"Committing with this message: {message}")
```
Comment (legoktm) [inline securedrop/i18n_tool.py:462]: Mostly just curious if there was a reason you added debug logging here.

**Review by legoktm — CHANGES_REQUESTED** [review]
This is really great! One question inline but otherwise the code looks good. I just merged #6562, which removed the `sh.sed` calls.

Could I ask you to rebase this on top of `develop`? Then we can confirm that all the `sh` imports are gone. (If you're not comfortable with git-rebase, I can help out with that part)

And once they're gone, we can remove it from `securedrop-app-code-requirements.in` and run `make update-python3-requirements` to remove it from `securedrop-app-code-requirements.txt`.

I appreciate you making it super easy to review by splitting it into very fine grained commits, just a heads up I'll squash it down before merging, so if you want to squash it yourself that's fine (and then "Verified" signature works properly).

**Comment by legoktm** [comment]
Would `subprocess.check_call(...)` be a better equivalent than `subprocess.run()` since we do want to throw an exception/halt when the underlying command exits with an error? Or we explicitly pass `run(..., check=True)` to every callsite.

File: securedrop/i18n_tool.py (hunk 1 above)
Comment (lb803) [reply securedrop/i18n_tool.py:462]: Sure! 
This assertion in the test suite
https://github.com/freedomofpress/securedrop/blob/3036ebbf9d0b49e2b3ad82851b0a38b30b8c2c43/securedrop/tests/test_i18n_tool.py#L312
checks that the stdout of the `git("-C", args.root, "commit", "-m", message, *paths)` command is present in a log (specifically, an object called caplog).

This confused me a bit, but I concluded that the git command I was replacing was piping the stdout to the log. I couldn't think of way to reproduce this behavior with subprocess, so I simply added a new log.debug statement.

Is this OK? Am I missing something?

**Comment by lb803** [comment]
Thanks @legoktm for your feedback! I'll work on this tomorrow. 

If it's OK for me to ask, I have a couple of questions: 

- Going forward, in which scenarios should I group several changes into a single commit (and so removing the need of "squashing them down" once a PR is submitted)?
- I need to learn more about mypy, but the output of the CI linter is a bit puzzling (mostly because it picks on three statements and overlooks many other -identical- instances). Where should I start looking to solve this?
```
securedrop/i18n_tool.py:275: error: No overload variant of "run" matches argument types "List[str]", "Dict[str, str]"
securedrop/i18n_tool.py:275: note: Possible overload variants:
securedrop/i18n_tool.py:275: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., input: Optional[str] = ..., text: Literal[True], timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:275: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: str, errors: Optional[str] = ..., input: Optional[str] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:275: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: str, input: Optional[str] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:275: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., *, universal_newlines: Literal[True], startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., input: Optional[str] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:275: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: Literal[False] = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: None = ..., errors: None = ..., input: Optional[bytes] = ..., text: Optional[Literal[False]] = ..., timeout: Optional[float] = ...) -> CompletedProcess[bytes]
securedrop/i18n_tool.py:275: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., input: Optional[Union[bytes, str]] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[Any]
securedrop/i18n_tool.py:277: error: No overload variant of "run" matches argument types "List[str]", "Dict[str, str]"
securedrop/i18n_tool.py:277: note: Possible overload variants:
securedrop/i18n_tool.py:277: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., input: Optional[str] = ..., text: Literal[True], timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:277: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: str, errors: Optional[str] = ..., input: Optional[str] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:277: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: str, input: Optional[str] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:277: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., *, universal_newlines: Literal[True], startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., input: Optional[str] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[str]
securedrop/i18n_tool.py:277: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: Literal[False] = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: None = ..., errors: None = ..., input: Optional[bytes] = ..., text: Optional[Literal[False]] = ..., timeout: Optional[float] = ...) -> CompletedProcess[bytes]
securedrop/i18n_tool.py:277: note:     def run(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stdout: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, capture_output: bool = ..., check: bool = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., input: Optional[Union[bytes, str]] = ..., text: Optional[bool] = ..., timeout: Optional[float] = ...) -> CompletedProcess[Any]
securedrop/i18n_tool.py:279: error: No overload variant of "check_output" matches argument types "List[str]", "str", "Dict[str, str]"
securedrop/i18n_tool.py:279: note: Possible overload variants:
securedrop/i18n_tool.py:279: note:     def check_output(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., text: Literal[True]) -> str
securedrop/i18n_tool.py:279: note:     def check_output(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: str, errors: Optional[str] = ..., text: Optional[bool] = ...) -> str
securedrop/i18n_tool.py:279: note:     def check_output(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: str, text: Optional[bool] = ...) -> str
securedrop/i18n_tool.py:279: note:     def check_output(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., *, universal_newlines: Literal[True], startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., text: Optional[bool] = ...) -> str
securedrop/i18n_tool.py:279: note:     def check_output(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: Literal[False] = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: None = ..., errors: None = ..., text: Optional[Literal[False]] = ...) -> bytes
securedrop/i18n_tool.py:279: note:     def check_output(args: Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., text: Optional[bool] = ...) -> Any
```

File: securedrop/i18n_tool.py (hunk 1 above)
Comment (legoktm) [reply securedrop/i18n_tool.py:462]: Ahh, thanks for the explanation. LGTM!

**Comment by legoktm** [comment]
> * Going forward, in which scenarios should I group several changes into a single commit (and so removing the need of "squashing them down" once a PR is submitted)?

We don't really have strict rules, in this I think you're making the same change across 2 files, when we want to look back in git history having it all in one commit seems sufficient. I usually split into multiple commits when I'm making multiple sets of distinct-ish changes that have logical separation.

> * I need to learn more about mypy, but the output of the CI linter is a bit puzzling (mostly because it picks on three statements and overlooks many other -identical- instances). Where should I start looking to solve this?

I am very puzzled too so I'm going to poke at it as well. Typically I just add in `reveal_type()` (see https://mypy.readthedocs.io/en/latest/common_issues.html#reveal-type) in a bunch of places until it makes sense.

**Comment by legoktm** [comment]
Oh and I was playing with something else earlier and noticed that your PR fixed two of the pylint checks we have suppressed in `Makefile`, unexpected-keyword-arg & too-many-function-args if you want to remove those disables :-)

**Comment by legoktm** [comment]
In `update_docs()`, our `k = {"cwd": includes}` gets typed as `builtins.dict[builtins.str*, builtins.str*]`. mypy is not smart enough to see that `cwd` is a literal parameter, since it doesn't see arbitrary kwargs in the type signature it complains.

But much earlier in `ensure_i18n_remote()` we have a `k = {"cwd": args.root}` that doesn't complain. This is typed as `builtins.dict[builtins.str*, Any]` and I think because it's "Any", it gets to skip some type checks or is already included in the subprocess.run/check_output type signatures.

My suggestion would be to just pass cwd as explicit parameters instead of kwargs. Yes, it's a little bit of repetition, but it's very minor and unlikely to constantly be changing that it'll be fine.

**Comment by lb803** [comment]
Fantastic, thank you so much for your comments!! @legoktm

**Review by legoktm — APPROVED** [review]
Nice, LGTM! Thanks and congrats for your first patch to SecureDrop! 🚀

**Pull request merged**